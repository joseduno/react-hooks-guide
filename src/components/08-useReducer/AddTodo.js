import React from 'react'
import useForm from '../../hooks/useForm';


const AddTodo = ({handleAdd}) => {

    const [{todo}, handleChange, reset] = useForm({todo: ''});

    const handleSubmit = e => {
        e.preventDefault()
        if (todo.trim().length <= 2) {
            return;
        }
        const newTodo = {
            id: new Date().getTime(),
            desc: todo,
            done: false
        }
        handleAdd(newTodo)
        reset()
    }

    return (
        <>
            <h4>Agregar TODO</h4>
            <hr />
            <form onSubmit={handleSubmit}>
                <input 
                    type='text'
                    name='todo'
                    className='form-control form-control-sm'
                    placeholder='Nueva tarea...'
                    autoComplete='off'
                    onChange={handleChange}
                    value={todo}
                />
                <button type='submit' className='btn btn-outline-primary mt-2 btn-sm w-100'>Agregar</button>
            </form>
        </>
    )
}

export default AddTodo
