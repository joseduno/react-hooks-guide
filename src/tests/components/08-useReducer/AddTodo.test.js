import AddTodo from "../../../components/08-useReducer/AddTodo"
import { shallow } from "enzyme";

describe('Pruebas en componente <AddTodo />', () => {

    const handleAdd = jest.fn()
    const wrapper = shallow(<AddTodo handleAdd={handleAdd} />)

    test('Debe mostrarse correctamente', () => {
        expect(wrapper).toMatchSnapshot()
    })

    test('no debe llamar handleAdd', () => {
        wrapper.find('form').simulate('submit', {preventDefault(){}})
        expect(handleAdd).not.toHaveBeenCalled()
    })

    test('debe llamar handleAdd', () => {
        const value = 'Aprender Express'

        wrapper.find('input').simulate('change', {target: {
            value,  // esto es el value del input
            name: 'todo'  // esto es el name del input
        }})
        wrapper.find('form').simulate('submit', {preventDefault(){}})
        expect(handleAdd).toHaveBeenCalled()
        expect(handleAdd).toHaveBeenCalledWith(expect.any(Object))
        expect(handleAdd).toHaveBeenCalledWith({
            id: expect.any(Number),  // cualquier numero
            desc: value,
            done: false
        })
        expect(wrapper.find('input').prop('value')).toBe('')
    })
    
    
})
