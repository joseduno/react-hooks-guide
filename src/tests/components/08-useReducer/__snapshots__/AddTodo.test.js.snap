// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en componente <AddTodo /> Debe mostrarse correctamente 1`] = `
<Fragment>
  <h4>
    Agregar TODO
  </h4>
  <hr />
  <form
    onSubmit={[Function]}
  >
    <input
      autoComplete="off"
      className="form-control form-control-sm"
      name="todo"
      onChange={[Function]}
      placeholder="Nueva tarea..."
      type="text"
      value=""
    />
    <button
      className="btn btn-outline-primary mt-2 btn-sm w-100"
      type="submit"
    >
      Agregar
    </button>
  </form>
</Fragment>
`;
